math/optimized_trig.h

```diff
@@ -0,0 +1,233 @@
+   const int32_t FP_PI_HALVES_Q = FP_PI_Q>>1; // qQ
+   const int32_t FP_PI = (1<<qX);  // qX
+   const int32_t FP_PI_HALVES = FP_PI>>1;  // qX
+
+   const double G=M_PI/2;
+   const double G2=G*G;
+   const double G3=G2*G;
+   const double G4=G2*G2;
+   const double G6=G4*G2;
+
+   // a - x^2*((b - x^2*(c - d*x^2))
+   // Coefficients found by minimizing the RMS error, and ensuring that the sine wave passes perfectly by 0 and 1.
+   const double a=1;
+   const double b = (-45045/G4*(G6+2*G4-156*G2+360)/(G3) - (-655*G-4504))/(952*G2); //# 4.933160419860462,;  // 0.499833652837794
+   const double c = 1/(2*G4)*(G - 6 + 4*b*G2);  // 0.041388888146338884
+   const double d = 1/(G6)*(G/2 - 2 + b*G2);  // 0.0012440842722145661
+
+   // Transform x to q, where q = x/pi
+   // aq - q^2*((bq - q^2*(cq - dq*q^2))
+   const double aq = a;
+   const double bq = b * G2*4;
+   const double cq = c * G4*16;
+   const double dq = d * G6*64;
+
+   // Convert the coefficients to fixed point, and inject a negative sign for certain ones. This negative
+   // sign will be useful when using the multiply & accumulate assembly instruction.
+   const int32_t AQ = aq * (1UL<<qN);  // qN
+   const int32_t BQm = -(bq * (1UL<<qN) + 0.5L);  // qN
+   const int32_t CQ = cq * (1UL<<qN) + 0.5L;  // qN
+   const int32_t DQm = -(dq * (1UL<<((48-(32-qN)-2*qX))) + 0.5L);  // ((16 + qN - 2*qX)))
+
+   // Convert the angle to fixed-point.
+   int32_t angle_FP = (angle_rad*PI_INVERSE)  * (1<<qQ); //  Do not round, as it is important not to reach +2^15
+
+   // By checking thebit in the PI place, we can determine if the angle is in the 3rd or 4th quadrant
+   bool is3rdOr4thQuadrant = angle_FP & FP_PI_Q;
+
+   // By checking whether there is a bit in the PI/2 place, we can determine if the angle is in the 2nd or 4th quadrant
+   bool is2ndOr4thQuadrant = angle_FP & FP_PI_HALVES_Q;
+
+   // Use the absolute value as cos(x) = cos(-x). Use a mask to truncate the value between 0 and pi.
+   int angle_truncated_FP = abs(angle_FP) & (FP_PI_Q-1);  // qQ
+
+   // Handle the special case of n*pi/2, which causes trouble for integer overflow
+   if (angle_truncated_FP == FP_PI_HALVES_Q) {
+      // Sine(+-pi/2) changes sign based on the quadrant
+      if (is3rdOr4thQuadrant == true) {
+         *sin_out = -1;
+      } else {
+         *sin_out = 1;
+      }
+
+      *cos_out = 0;
+      return;
+   }
+
+   // Use symmetry to calculate across [pi/2, pi). This is handy because the taylor coefficients are optimized for in this range.
+   const int32_t Q = angle_truncated_FP < FP_PI_HALVES_Q ? angle_truncated_FP : FP_PI_Q - angle_truncated_FP;  // qQ
+
+   // Shift X for proper scaling for use in the fixed-point math
+   int32_t X = Q >> (qQ-qX);  //qX
+
+   // Cosine(x). This MUST be done first
+   {
+      int32_t output;  // qN
+
+      // If the angle is close to pi/2, then use the simple approximation cos(pi/2-eps) ~= pi/2-eps
+      if (Q > (int32_t)(FP_PI_HALVES_Q*.95)) {
```

Comment:
It will be great to use cutoff where accuracy os simple and 6th order approximation meet. Also take care to avoid discontinuity at this point.
Also different running time may theoretically cause nasty bugs if calling code timing is important.

Reply:
> It will be great to use cutoff where accuracy os simple and 6th order approximation meet. Also take care to avoid discontinuity at this point.

The cutoff is around there. I didn't explore the space carefully, but it's around the 95% mark.

I had thought about redoing my coefficient calculation in order to avoid potential discontinuity, but ultimately decided it wasn't important because the discontinuity in C(0) and C(1) is small compared to the quantization errors on that scale.

> Also different running time may theoretically cause nasty bugs if calling code timing is important.

I am somewhat wary about that as well. It's easy to see how optimization could lead developers to believe there is more free CPU than is safe to use, ultimately resulting in a 1202 problem. But it feels unlikely in this case, since the optimization will only affect, at most, one of the two sine(x)/cos(x) operations and so the bulk of the function still runs no matter what.